Fb provenance runbuilder 6516

labkey-chrisj opened the pull request:
This PR adds components that were written to support provenance RunBuilder testing, but which belong in testAutomation

labkey-chrisj committed:
Merge branch 'release19.3-SNAPSHOT' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Item #6596
- add support for tab selection with tabbed grids

labkey-chrisj committed:
Item #6622
- add rbtCombo to support Provenance

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_provenance_runbuilder_6516

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_provenance_runbuilder_6516

labkey-chrisj committed:
Item #6622
- extract grid-tab support into new subcomponents; GridTab and GridTabBar

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_provenance_runbuilder_6516

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_provenance_runbuilder_6516

labkey-chrisj committed:
cr feedback for provenance testing
- fix finder for rbtCombo
- add file writer to WebDavUploadHelper

labkey-chrisj committed:
cr feedback for provenance runBuilder testing
- move FileUploadPanel, FileSelectTree, CollapsableFormPanel into testAutomation

labkey-chrisj committed:
cr feedback for provenance runBuilder testing
- rename CollapsiblePanel

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/RbtCombo.java:
```diff
@@ -0,0 +1,133 @@
+
+import java.util.List;
+
+public class RbtCombo extends WebDriverComponent<RbtCombo.ElementCache>
```
I don't think this is a LabKey developed component. Also, this name doesn't help people find it or know what it is for. Looking at the product code, I think this is a [`react-bootstrap-typeahead`](http://ericgio.github.io/react-bootstrap-typeahead/) component.
Move/rename this to `org.labkey.test.components.react.Typeahead` and add a class javadoc indicating what it represents. (Here's where the component is imported in the run builder: https://github.com/LabKey/provenance/blob/0e9c1e3e8046e16eb600bc8079dc160ff2238b6b/src/client/RunBuilder/PropertyLookup.tsx#L3)

@labkey-martyp Can you confirm?

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/RbtCombo.java:
```diff
@@ -0,0 +1,133 @@
+    public boolean isOpened()
+    {
+        return elementCache().input.getAttribute("aria-expanded").equals("true");
+    }
```
Be mindful of the API surface of components, especially in core test components.
Make this `private` or `protected`

labkey-chrisj replied:
I'd considered the cost/benefit of exposing it or not, then I figured it didn't hurt to let the test have access to it/seemed silly to have the functionality but not let a test access it

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_provenance_runbuilder_6516

labkey-chrisj committed:
cr feedback for provenance runBuilder testing
- rename TypeAheadCombo, add retry/synchronization to file upload input

labkey-chrisj committed:
cr feedback for provenance runBuilder testing
- hope just waiting a second works

labkey-danield commented on src/org/labkey/test/components/glassLibrary/CollapsiblePanel.java:
```diff
@@ -0,0 +1,119 @@
+    public boolean isExpanded()
+    {
+        return elementCache().waitForBody()
+                .getAttribute("class").equals("panel-collapse collapse in");
```
I don't think that is the best way to do this check. If the body is missing this will throw. Also It is dependent upon order of the attributes in the class.

I think it would be better to check for the existence of the  <div class="panel-body"> element in the panel. If it is not there the panel is not expanded.

labkey-chrisj replied:
If the body isn't there, the question of whether or not it is expanded is sorta meaningless

I will do this check by looking at the expand/collapse toggle (showing a minus-square means it's expanded, showing a plus-square means it's collapsed), but any way we look at it, we will throw if there's no body element

labkey-danield commented on src/org/labkey/test/components/glassLibrary/CollapsiblePanel.java:
```diff
@@ -0,0 +1,119 @@
+                .findWhenNeeded(this).withTimeout(2000);
+        WebElement expandCollapseToggle = Locator.tagWithClass("span", "pull-right")
+                .findWhenNeeded(header).withTimeout(2000);
+        private Locator.XPathLocator body = Locator.tagWithClass("div", "panel-collapse");
```
I don't think the body tag is correct the body is the <div class="panel-body"> tag contained in the domain-form-panel and the panel-collapse div:
<div class="domain-form-panel">
    <div class="panel-collapse">
        <div class="panel-body">
What you have will work, but you may want to avoid confusion in the future and reference the div tag that has the class value of panel-body.

labkey-danield commented on src/org/labkey/test/components/glassLibrary/CollapsiblePanel.java:
```diff
@@ -0,0 +1,119 @@
+
+        WebElement waitForBody()
+        {
+            return body.waitForElement(this, WAIT_FOR_JAVASCRIPT);
```
I know @labkey-tchad suggested you wait for the element in the cache, I think we should talk as a team and come up with a standard on when to wait for elements.

labkey-danield commented on src/org/labkey/test/components/glassLibrary/CollapsiblePanel.java:
```diff
@@ -0,0 +1,119 @@
+                .findWhenNeeded(this).withTimeout(2000);
+        WebElement expandCollapseToggle = Locator.tagWithClass("span", "pull-right")
+                .findWhenNeeded(header).withTimeout(2000);
+        private Locator.XPathLocator body = Locator.tagWithClass("div", "panel-collapse");
```
I don't think the body tag is correct the body is the <div class="panel-body"> tag contained in the domain-form-panel and the panel-collapse div:
<div class="domain-form-panel">
    <div class="panel-collapse">
        <div class="panel-body">
What you have will work, but you may want to avoid confusion in the future and reference the div tag that has the class value of panel-body.

labkey-chrisj replied:
I used the panel-collapse element (parent to the panel-body element) for 2 reasons: 
1) its class is used to signal its expand/collapse state
2) the use case I'm looking at puts provenance/runbuilder-specific classes in the panel-body div, I decided not to assume that it would exist in other usage

If we want to make that assumption, okay

labkey-tchad reviewed (approved):
Provenance tests passed.
https://teamcity.labkey.org/viewLog.html?buildId=923770

labkey-danield commented on src/org/labkey/test/components/glassLibrary/CollapsiblePanel.java:
```diff
@@ -0,0 +1,119 @@
+        {
+            return body.waitForElement(this, WAIT_FOR_JAVASCRIPT);
+        }
+    }
```
A couple other elements that would be useful for a panel:
Panels can have an error message that appears right below them. Things like missing values in required fields can cause them. It would be useful if we could get the error message associated with the panel.
Second the the icon on the left hand side changes, it would be useful tog et that to validate that the expected icon is shown.
Third a panel can have text on the right hand side next to the expand collapse button. For example this gives the status of the fields in the panel. it would be useful to grab that text as well.

labkey-danield commented on src/org/labkey/test/components/glassLibrary/components/FileUploadPanel.java:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        Locator.XPathLocator attachedFileContainer = Locator.tagWithClass("div", "attached-file--container")
+                .withChild(Locator.tagWithClass("span", "fa-times-circle"));
```
Just curious. Is the model now to put locators in the ElementCache and not in a Locators class?

labkey-chrisj replied:
I think yes- locators can go in either, which sorta makes the static Locators class redundant
But, I don't have strong feelings either way

labkey-danield commented on src/org/labkey/test/components/glassLibrary/components/react/typeahead/TypeAheadCombo.java:
```diff
@@ -0,0 +1,136 @@
+
+        elementCache().option(value).click();
+        WebDriverWrapper.waitFor(()-> !isOpened() && getValue().equals(value),
+                "the value was not set, or the combo did not close", 2000);
```
I like that you check that the value was set in the control.

labkey-danield commented on src/org/labkey/test/components/glassLibrary/components/react/typeahead/TypeAheadCombo.java:
```diff
@@ -0,0 +1,136 @@
+        return this;
+    }
+
+
```
Is there a way to get the values that are in the drop down list of the combo? A security test for example may need to check that only specific values are shown in a list/combo control to users with limited permissions.

labkey-chrisj replied:
They will appear as rendered dom elements in the flyout; should be relatively easy to implement

labkey-danield reviewed (changes requested):
Unfortunately I only had time to review a couple of the files.
I would suggest that you put a try catch around the isExpanded check for the panel. It will throw if the body is not there.
All the other comments/suggestions can wait until later.

labkey-martyp committed:
Add wait for file tree navigation

- First time opening a directory will display a Loading... spinner while querying the server.  Wait for this to go away.

labkey-chrisj committed:
cr feedback for provenance runBuilder testing
- use toggle button instead of body to check expanded/collapsed state

labkey-martyp committed:
Merge branch 'develop' into fb_provenance_runbuilder_6516
